### PR TITLE
[Merged by Bors] - small ecs cleanup and remove_bundle drop bugfix

### PR DESCRIPTION
**Pull request by Frizi**
- simplified code around archetype generations a little bit, as the special case value is not actually needed
- removed unnecessary UnsafeCell around pointer value that is never updated through shared references
- fixed and added a test for correct drop behaviour when removing sparse components through remove_bundle command

### CODE REVIEW

File: crates/bevy_ecs/src/archetype.rs [hunk 1]
```diff
@@ -317,8 +317,8 @@ pub struct ArchetypeGeneration(usize);
 
 impl ArchetypeGeneration {
     #[inline]
-    pub fn new(generation: usize) -> Self {
-        ArchetypeGeneration(generation)
+    pub fn initial() -> Self {
+        ArchetypeGeneration(0)
```
Comment (NathanSWard) [inline crates/bevy_ecs/src/archetype.rs:321]: Is there a reason we're using 0 instead of `usize::MAX` like the previous implementation did?

Also the function could be marked `const`

File: crates/bevy_ecs/src/system/commands.rs [hunk 2]
```diff
@@ -417,9 +417,29 @@ impl<T: Component> Command for RemoveResource<T> {
...
+    impl DropCk {
+        fn new_pair() -> (Self, Arc<AtomicUsize>) {
+            let atomic = Arc::new(AtomicUsize::new(0));
+            (DropCk(atomic.clone()), atomic)
+        }
+    }
+
+    impl Drop for DropCk {
+        fn drop(&mut self) {
+            self.0.as_ref().fetch_add(1, Ordering::AcqRel);
```
Comment (NathanSWard) [inline crates/bevy_ecs/src/system/commands.rs:440]: Since these are just tests and not performance critical, it's probably best to use `SeqCst` to avoid an possible weird atomic edge cases.

**Review by NathanSWard — CHANGES_REQUESTED** [review]

File: crates/bevy_ecs/src/archetype.rs (hunk 1 above)
Comment (Frizi) [reply crates/bevy_ecs/src/archetype.rs:321]: This is pretty much the point of this PR - simplify the initialization logic by not making MAX a special-case initial value, but use 0 value directly instead.

File: crates/bevy_ecs/src/system/commands.rs (hunk 2 above)
Comment (Frizi) [reply crates/bevy_ecs/src/system/commands.rs:440]: I mean, it's an atomic counter, it could really be Relaxed here, as there are no dependencies across other loads and stores. AcqRel is plenty already. SeqCst is only ever useful if you need to guarantee sequential consistency between multiple atomic loads and stores that involve multiple memory addresses, which is not the case here.

Comment (Frizi) [reply crates/bevy_ecs/src/system/commands.rs:440]: I mean, it's an atomic counter, it could really be Relaxed here, as there are no dependencies across other loads and stores. AcqRel is plenty already. SeqCst is only ever useful if you need to guarantee sequential consistency between multiple atomic loads and stores that involve multiple memory addresses, which is not the case here.

Comment (NathanSWard) [reply crates/bevy_ecs/src/system/commands.rs:440]: Yea I was just confused as to why it was `AcqRel` since it doesn't really fit here. 
Generally I push for defaulting to `SeqCst` and then case by case choosing a differing ordering. So I would argue for chosing `Relaxed` since it makes the most sense here.

Comment (Frizi) [reply crates/bevy_ecs/src/system/commands.rs]: Changed to `Relaxed`.

File: crates/bevy_ecs/src/world/entity_ref.rs [hunk 3]
```diff
@@ -586,13 +587,18 @@ unsafe fn get_component_and_ticks(
...
+///
 /// # Safety
-// `entity_location` must be within bounds of the given archetype and `entity` must exist inside the
-// archetype
-/// The relevant table row must be removed separately
-/// `component_id` must be valid
+/// - `entity_location` must be within bounds of the given archetype and `entity` must exist inside the archetype
+/// - `component_id` must be valid
+/// - The relevant table row **must be removed** by the caller once all components are taken
+/// - caller is responsible to drop component data behind returned pointer
```
Comment (bjorn3) [inline crates/bevy_ecs/src/world/entity_ref.rs:599]: It is safe to forget to drop it, so it shouldn't be under safety.

Comment (Frizi) [reply crates/bevy_ecs/src/world/entity_ref.rs]: Right. Moved it above safety, as it is still important note for correctly behaving code.

**Comment by cart** [comment]
This looks good to me. Love the archetype generation simplification!

I added UnsafeCell largely because [hecs does it in a similar context](https://github.com/Ralith/hecs/blob/master/src/archetype.rs#L35). We never create shared mutable references to a particular "range" in `data`, but we _do_ access different offsets of `data` mutably at the same time (ex: we allow accessing entity1's component data at the same time we access entity2's component data). We also allow shared references of Tables (which use BlobVec) to exist at the same time that mutable references to `data` exist (which I believe is what the hecs comment calls out).

As long as these use cases are ok, I'm down to remove it. But I want to be doubly sure before removing it (and im not even singly sure at this point :smile:).

**Comment by Frizi** [comment]
`NonNull<T>` already represents `*mut T` with extra guarantee of not being a null pointer, no need to put it inside `UnsafeCell`. It is only required to have `UnsafeCell` in order to get `&mut T` out of `&T` in any way, but in our case there are no shared references involved, as the memory comes directly from `alloc` call.

**Comment by cart** [comment]
I buy that. I think that means that hecs could remove it too? @Ralith you might want to look in to this?

**Comment by Ralith** [comment]
Thanks for the ping! My knee-jerk reaction is that the interpretation implied by this change makes more sense; the argument for `UnsafeCell` to be used here always seemed suspiciously magical to me, though I ultimately played it safe. I'll run the question by experts before removing it, though.

File: crates/bevy_ecs/src/archetype.rs [hunk 4]
```diff
@@ -317,8 +317,8 @@ pub struct ArchetypeGeneration(usize);
 
 impl ArchetypeGeneration {
     #[inline]
-    pub fn new(generation: usize) -> Self {
-        ArchetypeGeneration(generation)
+    pub fn initial() -> Self {
```
Comment (NathanSWard) [inline crates/bevy_ecs/src/archetype.rs:320]:
```suggestion
    pub const fn initial() -> Self {
```

**Comment by Ralith** [comment]
Every source I've checked confirms that `UnsafeCell` is unnecessary here.

File: crates/bevy_ecs/src/storage/blob_vec.rs [hunk 5]
```diff
@@ -9,17 +8,17 @@ pub struct BlobVec {
     item_layout: Layout,
     capacity: usize,
     len: usize,
-    data: UnsafeCell<NonNull<u8>>,
-    swap_scratch: UnsafeCell<NonNull<u8>>,
+    data: NonNull<u8>,
+    swap_scratch: NonNull<u8>,
```
Comment (bjorn3) [inline crates/bevy_ecs/src/storage/blob_vec.rs:12]: I can confirm this is unnecessary. Neither pointer is itself changed behind `&self`. Only the pointed to value can be changed using `&self`, which is fine as they are raw pointers that are not derived from a reference.

**Comment by cart** [comment]
Sounds like we have consensus. A big thanks to everyone for providing insight!
